### ISSUE TITLE
An item with the same key has already been added. Key: 32

### ISSUE THREAD

**Issue by ClausElmann** [EgorBo/Toasts.Forms.Plugin#137]
Hi
I get this error quite a lot, but can't figure our why.

It is latest version, and latest Xamarin.Forms

System.ArgumentExceptionAn item with the same key has already been added. Key: 32

Raw
System.Collections.Generic.Dictionary<TKey, TValue>.TryInsert(TKey key, TValue value, InsertionBehavior behavior)<30030eb7f2fa418cbb609e3c70603269>:0
System.Collections.Generic.Dictionary<TKey, TValue>.Add(TKey key, TValue value)<30030eb7f2fa418cbb609e3c70603269>:0
Plugin.Toasts.NotificationBuilder.TimerFinished(Activity activity, string id, bool cancel)<b81182804c7e489493d406b6d34a5576>:0
Plugin.Toasts.NotificationBuilder.<>c__DisplayClass14_0.<Notify>b__0(object x)<b81182804c7e489493d406b6d34a5576>:0
System.Threading.Timer.Scheduler.TimerCB(object o)<30030eb7f2fa418cbb609e3c70603269>:0
System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()<30030eb7f2fa418cbb609e3c70603269>:0
System.Threading.ThreadPoolWorkQueue.Dispatch()<30030eb7f2fa418cbb609e3c70603269>:0
System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()<30030eb7f2fa418cbb609e3c70603269>:0

Called via:

public static Task<INotificationResult> SendToast(string description, string title = "Besked:")
        {
            var tcs = new TaskCompletionSource<INotificationResult>();

            Device.BeginInvokeOnMainThread(async () =>
            {
                try
                {
                    var notificator = DependencyService.Get<IToastNotificator>();
                    var result = await notificator.Notify(new NotificationOptions()
                    {
                        Title = title,
                        Description = description,
                        //IsClickable = true,
                        //WindowsOptions = new WindowsOptions() { LogoUri = "icon.png" },
                        ClearFromHistory = true,
                        //DelayUntil = DateTime.UtcNow.AddSeconds(10)
                    });
                    tcs.TrySetResult(result);
                }
                catch (Exception ex)
                {
                    tcs.TrySetException(ex);
                }
            });
            return tcs.Task;
        }

**Comment by adamped** [comment]
Fixed with 3.3.0-pre1

**Comment by softsan** [comment]
i am using 3.3.1 and i have same issue:


```
public   void Notify(ToastCode code, string title, string description)
        {

            try
            {

                var notificator = DependencyService.Get<IToastNotificator>();
                 
                var options = new NotificationOptions
                {
                    ClearFromHistory = true,
                    AllowTapInNotificationCenter = false,
                    WindowsOptions = new WindowsOptions() { LogoUri = "icon.png" },
                    Description = description,
                    AndroidOptions = new AndroidOptions()
                    {
                        HexColor =  "#36942c",
                        ForceOpenAppOnNotificationTap = true,
                    }
                };
                switch (code)
                {
                    case ToastCode.DEFAULT:
                        options.Title = "Info";
                        break;
                    case ToastCode.WARNING:
                        options.Title = "Warning";
                        break;
                    case ToastCode.ERROR:
                        options.Title = "Error";
                        break;
                     
                }
                var result = notificator.Notify(options);
            }
            catch (Exception ex)
            {
                //Catch exception
            }
        }
```


```
An item with the same key has already been added. Key: 0

at System.Collections.Generic.Dictionary`2[TKey,TValue].TryInsert (TKey key, TValue value, System.Collections.Generic.InsertionBehavior behavior) [0x000c1] in <528ee26963544028a38104b20d5220c4>:0 
  at System.Collections.Generic.Dictionary`2[TKey,TValue].Add (TKey key, TValue value) [0x00000] in <528ee26963544028a38104b20d5220c4>:0 
  at Plugin.Toasts.NotificationBuilder.Notify (Android.App.Activity activity, Plugin.Toasts.INotificationOptions options) [0x003ca] in <66be0979e8054e5980ab2195ed0244d5>:0 
  at Plugin.Toasts.ToastNotification+<>c__DisplayClass5_0.<Notify>b__0 () [0x00054] in <66be0979e8054e5980ab2195ed0244d5>:0 
  at System.Threading.Tasks.Task`1[TResult].InnerInvoke () [0x0000f] in <528ee26963544028a38104b20d5220c4>:0 
  at System.Threading.Tasks.Task.Execute () [0x00010] in <528ee26963544028a38104b20d5220c4>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <528ee26963544028a38104b20d5220c4>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <528ee26963544028a38104b20d5220c4>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <528ee26963544028a38104b20d5220c4>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <528ee26963544028a38104b20d5220c4>:0 
  at System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () [0x00000] in <528ee26963544028a38104b20d5220c4>:0 
  at Plugin.Toasts.ToastNotification+<Notify>d__5.MoveNext () [0x00076] in <66be0979e8054e5980ab2195ed0244d5>:0 
```

**Comment by adamped** [comment]
@softsan _ I don't have much time to look at this project currently. If you want to find out where the issue is and submit a PR, will be happy to merge.

**Comment by NVentimiglia** [comment]
@adamped 
I think the issue is here
https://github.com/EgorBo/Toasts.Forms.Plugin/blob/a4662785fe876595482ee46e54eb30f8cd7cb150/Toasts.Forms.Plugin.Droid/NotificationBuilder.cs#L239

We are adding a key to a dictionary, but not checking for a conflict. Per the stack trace, this looks about right.

That said, I noticed that this lock is not used. Perhaps we intended to lock the counter in the opening of the notify method? Feels very edgy that this method is called is such tight succession, perhaps something else is at play.

https://github.com/EgorBo/Toasts.Forms.Plugin/blob/a4662785fe876595482ee46e54eb30f8cd7cb150/Toasts.Forms.Plugin.Droid/NotificationBuilder.cs#L31

**Comment by adamped** [comment]
@softsan @NVentimiglia 

I had a quick look and I found the error. It has nothing to do with if the key actually existed, it is because the type Dictionary is not thread-safe, meaning if something was checking or changing the dictionary, while something was trying to add to it, it would cause this error.

I also added the lock around the setting of the Id, as it could possibly cause an issue, but I even set sending notification in a loop and I could never get the Id's out of order. But it was theoretically possible.

The fix is in 3.3.2:
https://www.nuget.org/packages/Toasts.Forms.Plugin/3.3.2

Also, @softsan via the code sample you send, you need to await the notify method. e.g.
`var result = **await** notificator.Notify(options);`

**Comment by NVentimiglia** [comment]
@adamped Thats great, thank you for your commitment!
